### PR TITLE
Add multi-host GPU support

### PR DESCRIPTION
**Pull request by vanbasten23**
Collaborating with @wbmc 

To start the multi-host (or multi-node) training, do:
```
PJRT_DEVICE=GPU torchrun \
--nproc_per_node=${NUMBER_LOCAL_GPU_DEVICES} \
--nnodes=${NUMBER_GPU_VM} \
--node_rank=${CURRENT_HOST_RANK} \
--rdzv_endpoint=<internal_ip_address> multinode_training_script.py
```
on each host.

The documentation will be updated in a follow-up PR.

### CODE REVIEW

**Comment by JackCaoG** [comment]
@wbmc I have granted you the write access

**Comment by wbmc** [comment]
> @wbmc I have granted you the write access

OK, Thanks!

File: torch_xla/csrc/init_python_bindings.cpp
Comment (wbmc) [inline torch_xla/csrc/init_python_bindings.cpp:1847]: The port number `8547` seems to be a hardcoded value, which could potentially lead to conflicts with other processes in a cloud environment. Is it possible to allow users to specify a different port number? Otherwise, they would need to recompile the torch_xla program.

File: torch_xla/csrc/runtime/pjrt_computation_client.cc
Comment (wbmc) [inline torch_xla/csrc/runtime/pjrt_computation_client.cc:135]: The port number `8547` seems to be a hardcoded value, which could potentially lead to conflicts with other processes in a cloud environment. Is it possible to allow users to specify a different port number? Otherwise, they would need to recompile the torch_xla program.

File: torch_xla/csrc/init_python_bindings.cpp
Comment (vanbasten23) [reply torch_xla/csrc/init_python_bindings.cpp:1847]: Yea, that makes sense. It should be possible.

Comment (vanbasten23) [reply torch_xla/csrc/init_python_bindings.cpp]: I added a flag COORDINATOR_PORT to set an alternative port to be used. In practice, I imagine the default port should be available most of the time. In case it isn't, users can unblock themselves that way.

File: torch_xla/csrc/runtime/pjrt_computation_client.cc
Comment (vanbasten23) [reply torch_xla/csrc/runtime/pjrt_computation_client.cc]: ditto

File: torch_xla/csrc/init_python_bindings.cpp
Comment (miladm) [inline torch_xla/csrc/init_python_bindings.cpp:1848]: For better readability, can we introduce a variable to describe this IP address is the default parameter? e.g/ `LOCAL_HOST_IP_DEFAULT`

Comment (vanbasten23) [reply torch_xla/csrc/init_python_bindings.cpp:1848]: sure

File: test/test_train_mp_imagenet_torchrun.py
Comment (vanbasten23) [inline test/test_train_mp_imagenet_torchrun.py:1]: to reviewers: this is a torchrun version of test/test_train_mp_imagenet.py. 
Maybe I should merge it with test/test_train_mp_imagenet.py..

File: test/pjrt/test_torchrun.py
Comment (jonb377) [inline test/pjrt/test_torchrun.py:35]: Can we init/destroy the process group in setUpTest/tearDownTest to cover the early exit case?

File: test/pjrt/test_torchrun.py
Comment (jonb377) [inline test/pjrt/test_torchrun.py:40]: Could you add a comment on what value the tensor should have? It's a bit hard to follow at the moment.

File: test/test_train_mp_imagenet_torchrun.py
Comment (jonb377) [reply test/test_train_mp_imagenet_torchrun.py:1]: Are there any differences?

File: torch_xla/_internal/gpu.py
Comment (jonb377) [inline torch_xla/_internal/gpu.py:36]: Neat! So this will shutdown the distributed runtime when the rank-0 process exits?

I'm curious if this process somehow finishes first, will the shutdown block until all clients are destroyed?

File: torch_xla/csrc/init_python_bindings.cpp
Comment (jonb377) [inline torch_xla/csrc/init_python_bindings.cpp:1847]: How about we make the address and port parameters to `_xla_get_distributed_runtime_service`? I'll need this shortly for autocheckpointing, so happy to pick up the refactor after you land this PR.

File: torch_xla/csrc/runtime/pjrt_computation_client.cc
Comment (jonb377) [inline torch_xla/csrc/runtime/pjrt_computation_client.cc:139]: How about `sys_util::GetEnvInt("RANK", local_rank);`, then we don't need the helper function

File: torch_xla/csrc/runtime/pjrt_computation_client.cc
Comment (jonb377) [inline torch_xla/csrc/runtime/pjrt_computation_client.cc:143]: We need to generalize this to support `CUDA_VISIBLE_DEVICES` and single-process-multi-device

File: torch_xla/csrc/init_python_bindings.cpp
Comment (jonb377) [inline torch_xla/csrc/init_python_bindings.cpp:1846]: Is `COORDINATOR_PORT` documented anywhere?

File: torch_xla/_internal/gpu.py
Comment (jonb377) [inline torch_xla/_internal/gpu.py:18]: Is this right? We'll clobber the torchrun-set world size. Also wondering if we need to keep `GPU_NUM_DEVICES` in the first place

File: torch_xla/_internal/pjrt.py
Comment (jonb377) [inline torch_xla/_internal/pjrt.py:123]: Also for consideration: `xu.getenv_as(xenv.RANK, int, local_rank)`

**Review by jonb377 — COMMENTED** [review]
Looking great Xiongfei!

File: test/test_train_mp_imagenet_torchrun.py
Comment (vanbasten23) [reply test/test_train_mp_imagenet_torchrun.py:1]: Only minor changes. In test_train_mp_imagenet.py, we have
```
if __name__ == '__main__':
  xmp.spawn(_mp_fn, args=(FLAGS,), nprocs=FLAGS.num_cores)
```
This file stops calling `spawn` and changes to:
```
if __name__ == '__main__':
  if not dist.is_torchelastic_launched():
    logging.error('Test must be launched with torchrun!')
    exit(1)
  _mp_fn(FLAGS)
```

File: torch_xla/csrc/runtime/pjrt_computation_client.cc
Comment (vanbasten23) [reply torch_xla/csrc/runtime/pjrt_computation_client.cc:143]: Yeah, we initially plan to incorporate CUDA_VISIBLE_DEVICES into this PR but we encountered some errors such as https://github.com/pytorch/xla/issues/5558#issuecomment-1756482280 and https://github.com/pytorch/xla/issues/5558#issuecomment-1751574292. We still plan to do it but probably in a follow-up pr

File: torch_xla/_internal/gpu.py
Comment (vanbasten23) [reply torch_xla/_internal/gpu.py:18]: This change is mainly to make single-host case work with `spawn` (e.g. the tests in pytorch/xla/test/pjrt/test_runtime_gpu.py in which we use `spawn`). To provide a similar UX as PyTorch, we should still support `spawn` for the single host case (fwiw, PyTorch supports it as https://screenshot.googleplex.com/7nKD68dXNUUskF7). But I like your idea of replacing "GPU_NUM_DEVICES" with "WORLD_SIZE".

Also, torchrun doesn't invoke `spawn` and in turn this function so it wouldn't clobber the torchrun-set world size.

So how about I replace `GPU_NUM_DEVICES ` with `WORLD_SIZE `?

File: test/test_train_mp_imagenet_torchrun.py
Comment (will-cromar) [reply test/test_train_mp_imagenet_torchrun.py:1]: Can you merge these two cases? ie if the program has been launched with torchrun, call `_mp_fn`, otherwise call `xmp.spawn`

File: torch_xla/_internal/gpu.py
Comment (will-cromar) [reply torch_xla/_internal/gpu.py:18]: `WORLD_SIZE` isn't quite right here, since this function returns the expected number of _local_ processes. In `torchrun`, that's set as `LOCAL_WORLD_SIZE`. In `xmp.spawn`, we had to set it as something different like `PJRT_LOCAL_WORLD_SIZE` because `LOCAL_WORLD_SIZE` caused some issue with a third party package.

So maybe `os.environ.get('LOCAL_WORLD_SIZE') or os.environ.get('PJRT_LOCAL_WORLD_SIZE')`? It's clunky, but it covers both cases.

To use a subset of local GPUs with `xmp.spawn`, a user could set then `LOCAL_WORLD_SIZE` themselves instead of `GPU_NUM_DEVICES`.

File: torch_xla/_internal/gpu.py
Comment (will-cromar) [reply torch_xla/_internal/gpu.py:36]: TIL this module exists. Cool!

doc for other reviewers who are curious https://docs.python.org/3/library/atexit.html

File: torch_xla/core/xla_env_vars.py
Comment (will-cromar) [inline torch_xla/core/xla_env_vars.py:29]: nit: are these constants provided by upstream `torch.distributed` somewhere?

File: torch_xla/csrc/init_python_bindings.cpp
Comment (will-cromar) [reply torch_xla/csrc/init_python_bindings.cpp:1846]: I think this name needs to be more verbose, unfortunately. There are too many coordinators in our system already: TPU slice discovery, multislice discovery, torch.distributed process groups, and `xla::DistributedRuntimeService`. Maybe `XLA_COORDINATOR_PORT` and `XLA_COORDINATOR_ADDR`?

File: torch_xla/_internal/pjrt.py
Comment (will-cromar) [inline torch_xla/_internal/pjrt.py:105]: This overlaps confusingly with the `global_ordinal` and the `process_index` in `torch_xla.runtime`, and this usage of the `RANK` env var or the local rank strikes me as a special case. `RANK` is not always set.

AFAICT, this index is used in one place: to determine whether to start the distributed runtime. How about something like `is_distributed_master(local_rank: int)` or `should_start_xla_coordinator(local_rank:int)` so we don't have another another index-getter?

cc @jonb377 for input since you're going to need to use this for checkpointing

File: torch_xla/csrc/runtime/pjrt_computation_client.cc
Comment (will-cromar) [inline torch_xla/csrc/runtime/pjrt_computation_client.cc:50]: nit: can you use `absl::StrJoin` instead? https://abseil.io/docs/cpp/guides/strings#abslstrjoin-for-joining-elements-within-a-string

File: torch_xla/csrc/runtime/pjrt_computation_client.cc
Comment (will-cromar) [reply torch_xla/csrc/runtime/pjrt_computation_client.cc:139]: It's time for me to bring up everyone's favorite topic: TPU v2/v3 multithreading :confetti_ball: 

Normally we could ignore this case entirely since we only needed the distributed runtime for GPUs. But, @jonb377 will be implementing it for TPUs as well, so we have to be more careful about the difference with process rank (shared between threads) and ordinal (unique to each thread) in this context.

Short term, can you make the naming more specific? ie `local_process_rank`, `global_process_rank`, and `getProcessRank` to emphasize these are process-scoped indices?

**Review by will-cromar — COMMENTED** [review]
Great work! I'm really excited to see your progress here.

While this PR is in review, can you start updating the documentation for GPUs in this repository? We should have some documentation covering how/why the GPU runtime works and make it clear that we expect people to use `torchrun` for multihost use cases.

File: torch_xla/_internal/pjrt.py
Comment (vanbasten23) [reply torch_xla/_internal/pjrt.py:105]: SG

File: torch_xla/_internal/pjrt.py
Comment (vanbasten23) [reply torch_xla/_internal/pjrt.py:123]: This is handy. Thanks!
I replaced it with a function `should_initialize_dist_runtime` per a previous comment.

File: torch_xla/core/xla_env_vars.py
Comment (vanbasten23) [reply torch_xla/core/xla_env_vars.py:29]: I did a search but couldn't find it. A lot of places in upstream uses the string literals.

File: torch_xla/csrc/init_python_bindings.cpp
Comment (vanbasten23) [reply torch_xla/csrc/init_python_bindings.cpp:1846]: > Is COORDINATOR_PORT documented anywhere?

COORDINATOR_PORT is used to overwrite the default port for pjrt's dist runtime service. I added it in response to another comment. Before it, we initialize the dist runtime service at the port 8547 by default. But when this port is not available (the possibility may be low I assume), users wouldn't be able to start the service and would have to recompile torch_xla. In that case, users can overwrite the default port and unblock themselves.

> Is there ever a case where the torch.distributed master will be different than the distributed runtime master?

So far, torch.distributed master and pjrt's dist runtime master can share the same address and use different ports without any issues. So I'll rename COORDINATOR_PORT to XLA_COORDINATOR_PORT.

File: torch_xla/csrc/runtime/pjrt_computation_client.cc
Comment (vanbasten23) [reply torch_xla/csrc/runtime/pjrt_computation_client.cc:50]: Thanks!

File: test/pjrt/test_torchrun.py
Comment (vanbasten23) [reply test/pjrt/test_torchrun.py]: done

File: test/pjrt/test_torchrun.py
Comment (vanbasten23) [reply test/pjrt/test_torchrun.py:46]: done

File: test/test_train_mp_imagenet_torchrun.py
Comment (vanbasten23) [reply test/test_train_mp_imagenet_torchrun.py]: done

File: torch_xla/_internal/pjrt.py
Comment (vanbasten23) [reply torch_xla/_internal/pjrt.py:106]: done

File: torch_xla/csrc/init_python_bindings.cpp
Comment (vanbasten23) [reply torch_xla/csrc/init_python_bindings.cpp]: done

File: torch_xla/csrc/runtime/pjrt_computation_client.cc
Comment (vanbasten23) [reply torch_xla/csrc/runtime/pjrt_computation_client.cc]: done

File: torch_xla/_internal/gpu.py
Comment (vanbasten23) [reply torch_xla/_internal/gpu.py:18]: Agreed that `WORLD_SIZE` is confusing. OTOH, `WORLD_SIZE` is what PyTorch single-node training uses: https://screenshot.googleplex.com/BH27HAYTpbNU4KA, if we want to stay closer to PyTorch.

I think clarity is more important here. How do you think we replace `GPU_NUM_DEVICES` with `PJRT_LOCAL_WORLD_SIZE` (or something like `PT_XLA_LOCAL_WORLD_SIZE` if we don't want to leak the underlying implementation detail), so the way we run single-host-multi-GPU `GPU_NUM_DEVICES=4 python3 xla/test/test_train_mp_imagenet.py ...` turns to `PJRT_LOCAL_WORLD_SIZE=4 python3 xla/test/test_train_mp_imagenet.p ..`?  @will-cromar @jonb377

File: torch_xla/csrc/init_python_bindings.cpp
Comment (vanbasten23) [reply torch_xla/csrc/init_python_bindings.cpp:1928]: per our discussion offline, let's push it down to the cpp layer

File: torch_xla/_internal/gpu.py
Comment (will-cromar) [reply torch_xla/_internal/gpu.py:18]: I think it's okay to rely on the `LOCAL_WORLD_SIZE` env var here. We know torchrun will set it for sure, and the manual single-host-multi-GPU case can become `PJRT_DEVICE=GPU LOCAL_WORLD_SIZE=4 python script.py`*. We can get rid of `GPU_NUM_DEVICES`.

\* As a follow up, I would like to implement more automatic configuration like we have with TPUs so users don't have to set anything in the default case

We definitely don't want to override `torchrun`'s settings here.

I forgot that `PJRT_LOCAL_WORLD_SIZE` is set after this function is called (and probably just set to the output of this function). So we can ignore that variable here.

File: torch_xla/csrc/runtime/pjrt_computation_client.cc
Comment (vanbasten23) [reply torch_xla/csrc/runtime/pjrt_computation_client.cc:163]: done

File: torch_xla/csrc/init_python_bindings.cpp
Comment (will-cromar) [inline torch_xla/csrc/init_python_bindings.cpp:1932]: nit: does this have to be an IP? Otherwise, you can just hardcode `"localhost"` as the default

File: torch_xla/csrc/init_python_bindings.cpp
Comment (will-cromar) [inline torch_xla/csrc/init_python_bindings.cpp:1935]: nit: update the error message to `MASTER_ADDR` too

File: torch_xla/_internal/pjrt.py
Comment (vanbasten23) [reply torch_xla/_internal/pjrt.py]: Yeah, it's gone in the latest iteration.

File: torch_xla/_internal/gpu.py
Comment (vanbasten23) [reply torch_xla/_internal/gpu.py]: Yeah, I used `LOCAL_WORLD_SIZE ` here.

I'll replace `GPU_NUM_DEVICES` with `LOCAL_WORLD_SIZE` in a follow-up PR.

File: torch_xla/csrc/init_python_bindings.cpp
Comment (vanbasten23) [reply torch_xla/csrc/init_python_bindings.cpp:1932]: yeah, "localhost" is better!

File: torch_xla/csrc/init_python_bindings.cpp
Comment (vanbasten23) [reply torch_xla/csrc/init_python_bindings.cpp]: done

File: torch_xla/csrc/init_python_bindings.cpp
Comment (jonb377) [reply torch_xla/csrc/init_python_bindings.cpp:1928]: Do you plan to take this in a follow-up PR? If so, I can drive it to unblock autocheckpointing, the GPU-related changes to the distributed runtime seem complete (feel free to let me know otherwise).

File: torch_xla/csrc/init_python_bindings.cpp
Comment (jonb377) [inline torch_xla/csrc/init_python_bindings.cpp:1931]: I'm concerned about hard-coding the default value in two places, can we make this a const somewhere? Or if you want, this can be addressed in the follow-up to move everything into C++.

File: torch_xla/csrc/runtime/pjrt_computation_client.cc
Comment (jonb377) [inline torch_xla/csrc/runtime/pjrt_computation_client.cc:155]: nit: Same here, splitting into two lines would be a bit cleaner.

**Review by jonb377 — APPROVED** [review]
LGTM, thanks Xiongfei!

File: torch_xla/_internal/pjrt.py
Comment (vanbasten23) [reply torch_xla/_internal/pjrt.py:125]: will resolve in a follow-up PR.

File: torch_xla/csrc/init_python_bindings.cpp
Comment (vanbasten23) [reply torch_xla/csrc/init_python_bindings.cpp:1928]: Yes, it's in https://github.com/pytorch/xla/pull/5714.

File: torch_xla/_internal/pjrt.py
Comment (vanbasten23) [reply torch_xla/_internal/pjrt.py:125]: it's in https://github.com/pytorch/xla/pull/5714

File: torch_xla/csrc/init_python_bindings.cpp
Comment (vanbasten23) [reply torch_xla/csrc/init_python_bindings.cpp:1931]: Fair. Addressed it in https://github.com/pytorch/xla/pull/5714

File: torch_xla/csrc/runtime/pjrt_computation_client.cc
Comment (vanbasten23) [reply torch_xla/csrc/runtime/pjrt_computation_client.cc:155]: done in https://github.com/pytorch/xla/pull/5714

**Pull request merged**
